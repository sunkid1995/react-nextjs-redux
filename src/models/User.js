import Joi from 'joi-browser';
import _ from 'lodash';

// Models
import BaseModel, { COMMON_SCHEMAS } from './BaseModel';

const { ID, STRING, STRINGS_ARRAY } = COMMON_SCHEMAS;

const ROLES = {
  MANAGER: 'sales_manager',
  STAFF: 'sales_man',
  USER: 'User',
};

const { MANAGER, STAFF } = ROLES;

export default class User extends BaseModel {
  static get klass() { return 'User'; }
  
  static dataValidator(): Joi.SchemaMap {
    return {
      auth_token: STRING.required(),
      display_name: STRING.required(),
      email: STRING.required(),
      id: ID,
      nickname: STRING.required(),
      picture: STRING,
      roles: STRINGS_ARRAY,
    };
  }

  get isManager() { return _.includes(this.roles, MANAGER); }
  get isStaff() { return _.includes(this.roles, STAFF) && !_.includes(this.roles, MANAGER); }
}
