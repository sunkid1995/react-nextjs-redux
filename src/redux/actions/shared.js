// import { createAction } from 'redux-actions';

// // Redux
// import { SHARED } from 'src/redux/types';
