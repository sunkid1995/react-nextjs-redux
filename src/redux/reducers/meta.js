import { REHYDRATE } from 'redux-persist/lib/constants';

const INITIAL_STATE = {
  rehydrated: false,
};

export default function metaReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === REHYDRATE) {
    const { err } = action;

    /**
     * On server-side, REHYDRATE action will throw an err:
     * Error: No available storage method found
     * Otherwise on client-side, REHYDRATE will be dispatched without error
     * So we check if err is nullable then it's actually rehydrated in client
     */
    if (err == null) return { ...state, rehydrated: true };
  }

  return state;
}
