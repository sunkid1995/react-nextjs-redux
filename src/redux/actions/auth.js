import { createAction } from 'redux-actions';

// Redux
import { AUTH } from 'src/redux/types';

export const authorize = createAction(AUTH.AUTHORIZE);
export const deauthorize = createAction(AUTH.DEAUTHORIZE);
