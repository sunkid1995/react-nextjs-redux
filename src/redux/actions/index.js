import * as auth from './auth';
import * as serviceApi from './serviceApi';
import * as shared from './shared';

export { auth, serviceApi, shared };
