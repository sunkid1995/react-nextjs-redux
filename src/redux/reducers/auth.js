// Redux
import { AUTH } from 'src/redux/types';

const INITIAL_STATE = {
  user: undefined,
};

export default function authReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === AUTH.AUTHORIZE) {
    const { payload: user } = action;
    return { ...state, user };
  }

  if (type === AUTH.DEAUTHORIZE) {
    return INITIAL_STATE;
  }

  return state;
}
