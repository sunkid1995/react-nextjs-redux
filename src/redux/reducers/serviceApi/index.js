import _ from 'lodash';

// Constants
import { AUTH, SERVICE_API } from 'src/redux/types';

// Locals
// import blankRequestReducer from './blankRequestReducer';
// import paginatedRequestReducer from './paginatedRequestReducer';
import plainRequestReducer from './plainRequestReducer';

const INITIAL_BLANK_REQUEST = { error: undefined, loading: false };
// const INITIAL_PAGINATED_DATA_REQUEST = { dataPages: {}, totalPages: 1, ...INITIAL_BLANK_REQUEST };
const INITIAL_PLAIN_DATA_REQUEST = { data: {}, ...INITIAL_BLANK_REQUEST };

const INITIAL_STATE = {
  emailAuth: INITIAL_PLAIN_DATA_REQUEST,
  passwordForgotten: INITIAL_PLAIN_DATA_REQUEST,
  passwordReset: INITIAL_PLAIN_DATA_REQUEST,
};

export default function serviceApiReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === AUTH.DEAUTHORIZE) return INITIAL_STATE;

  if (
    _.startsWith(type, SERVICE_API.AUTH_WITH_EMAIL) ||
    _.startsWith(type, SERVICE_API.FORGOT_PASSWORD) ||
    _.startsWith(type, SERVICE_API.RESET_PASSWORD)
  ) return plainRequestReducer(state, action);

  return state;
}
