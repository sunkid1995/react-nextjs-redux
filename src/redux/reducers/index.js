import auth from './auth';
import meta from './meta';
import serviceApi from './serviceApi';
import shared from './shared';

export { auth, meta, serviceApi, shared };
