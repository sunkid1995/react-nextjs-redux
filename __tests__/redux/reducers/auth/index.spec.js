import { auth as authActions } from 'src/redux/actions';
import { auth as authReducer } from 'src/redux/reducers';

// Helpers
import buildMockState from '__tests__/helpers/buildMockState';

// Models
import User from 'src/models/User';

describe('redux/reducers/auth', () => {
  const { auth: mockState } = buildMockState();
  const user = User.build(require('src/data/fixtures/user.json'));

  test('authorize', () => {
    const expectedState = { ...mockState, user };
    const state = authReducer(mockState, authActions.authorize(user));
    expect(state).toEqual(expectedState);
    expect(state).toMatchSnapshot();
  });

  test('deauthorize', () => {
    const currentState = { ...mockState, user };
    const state = authReducer(currentState, authActions.deauthorize());
    expect(state).toEqual(mockState);
    expect(state).toMatchSnapshot();
  });
});
