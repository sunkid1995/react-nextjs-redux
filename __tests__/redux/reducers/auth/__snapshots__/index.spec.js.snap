// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/reducers/auth authorize 1`] = `
Object {
  "user": User {
    "auth_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6InRoYW5obnhAb25hY2xvdmVyLmNvbSIsInVpZCI6IjgzOWU2OGJkLTJjN2QtNDkzMy1hMmI3LWYxZDM0MDkyODI0NiIsImlhdCI6MTUyMDg0MjA3MCwiZXhwIjoxNTIxNDQ2ODcwLCJqdGkiOiJiMDlhYjM4NS1iZWZiLTRiNzYtODQyZC0xODQ1ZWU2MGJhNGEifQ.zUxEMCY2zYj6y4jw6-sJAHig-hPzDY_l_MoCT4XfbhI",
    "display_name": "ThanhNX",
    "email": "thanhnx@onaclover.com",
    "id": 4,
    "nickname": "ThanhNX",
    "roles": Array [
      "user",
      "sales_manager",
      "sales_man",
    ],
  },
}
`;

exports[`redux/reducers/auth deauthorize 1`] = `
Object {
  "user": undefined,
}
`;
